### PR TITLE
fix: address feedback and fix noticed issued on move contact

### PR DESCRIPTION
**Pull request by paulpascal**
# Description:

We are encountering several issues with the MoveContactWorker that are causing job failures. These issues are related to token expiration, command timeouts, lack of error feedback, and server availability errors. 

This PR focus on addressing them

- [ ] Token Expiration #221:
Jobs that are retried (postponed) many times face a TokenExpiredError: jwt expired. This happens because the session token expires after 48 hours, but some jobs might take longer to complete, causing the token to expire before the job finishes. (This is quite what is intended, but what are we handling the  failed jobs in general ? manually rerun ? )

- [ ] Timeout Handling #221::
Some jobs are failing with the error cht-conf timed out. This occurs when the cht process takes too long to complete, and the job is prematurely terminated.

- [ ] Undefined Errors in cht Execution #221::
Occasionally, the cht process exits with a code 1 and displays ERROR undefined, making it difficult to diagnose the root cause of the failure. The logs provide limited information about why the command failed.

- [ ] Server Availability (HTTP 500):
The job occasionally encounters an HTTP 500 error while trying to connect to instances (e.g., https://kakamega-echis.health.go.ke/medic/). This could be due to server downtime or temporary server issues.

- [ ] Close Issue #222 - Include cht-conf Output in Job Logs:
To enhance transparency and facilitate easier troubleshooting, we should ensure that the output from cht-conf commands is included directly in the job logs. This will allow us to view critical information, such as errors from cht-conf, in one place for better visibility and faster diagnosis.

- [ ] Close Issue #184 - Improve Error Handling and Retry Logic for Interrupted/Failed Move Contact Jobs:
Enhance the error handling and retry logic for the move contact feature to ensure robustness in cases of connectivity loss or other interruptions. This would allow for automatic retries of failed jobs due to server connectivity issues or other interruptions, reducing manual intervention.

### CODE REVIEW

**Comment by paulpascal** [comment]
Hi @kennsippell, I have been able to work on some of the feedback we received from `move-contact` feature, would you have a look on this please?

Thanks

File: test/integration/move-contact.spec.ts [hunk 1]
```diff
@@ -116,7 +117,7 @@ describe('integration/move-contact',  function () {
     await new Promise(resolve => setTimeout(resolve, 1000));
     
     // Check if the job has failed
-    const job = await moveContactQueue['bullQueue'].getJob(jobId) as Job;
+    const job = await moveContactQueue['bullQueue'].getJob(jobId) as unknown as Job;
```
Comment (kennsippell) [inline test/integration/move-contact.spec.ts:120]: why needed?

File: src/worker/move-contact-worker.ts [hunk 2]
```diff
@@ -141,22 +164,28 @@ export class MoveContactWorker {
...
-    // Delayed this job by DELAY_IN_MILLIS, using the current worker processing token
-    await job.moveToDelayed(retryTime.toMillis(), processingToken);
+    return { retryTime, retryTimeFormatted };
+  }
 
-    const retryMessage = `Job ${job.id} postponed until ${retryTimeFormatted}.  Reason was sentinel backlog.`;
-    job.log(`[${new Date().toISOString()}]: ${retryMessage}`);
-    console.log(retryMessage);
+  private static jobLogWithTimestamp(job: Job|MinimalJob|undefined, message: string): void {
+    const timestamp = new Date().toISOString();
```
Comment (kennsippell) [inline src/worker/move-contact-worker.ts:186]: Stick with luxon in this file instead of mixing?

File: src/worker/move-contact-worker.ts [hunk 3]
```diff
@@ -40,40 +46,56 @@ export class MoveContactWorker {
...
+  };
+
+  private static async shouldPostpone(jobData: MoveContactData): Promise<{ shouldPostpone: boolean; reason: string }> {
     try {
       const { instanceUrl } = jobData;
       const response = await axios.get(`${instanceUrl}/api/v2/monitoring`);
       const sentinelBacklog = response.data.sentinel?.backlog;
       console.log(`Sentinel backlog at ${sentinelBacklog} of ${this.MAX_SENTINEL_BACKLOG}`);
-      return sentinelBacklog > this.MAX_SENTINEL_BACKLOG;
+      return { shouldPostpone: sentinelBacklog > this.MAX_SENTINEL_BACKLOG, reason: `Sentinel backlog too high at ${sentinelBacklog}` };
```
Comment (kennsippell) [inline src/worker/move-contact-worker.ts:82]:
```suggestion
      return {
        shouldPostpone: sentinelBacklog > this.MAX_SENTINEL_BACKLOG,
        reason: `Sentinel backlog too high at ${sentinelBacklog}`
      };
```

File: src/worker/move-contact-worker.ts [hunk 4]
```diff
@@ -141,22 +164,28 @@ export class MoveContactWorker {
...
-    
-    // Delayed this job by DELAY_IN_MILLIS, using the current worker processing token
-    await job.moveToDelayed(retryTime.toMillis(), processingToken);
+    return { retryTime, retryTimeFormatted };
+  }
 
-    const retryMessage = `Job ${job.id} postponed until ${retryTimeFormatted}.  Reason was sentinel backlog.`;
-    job.log(`[${new Date().toISOString()}]: ${retryMessage}`);
-    console.log(retryMessage);
+  private static jobLogWithTimestamp(job: Job|MinimalJob|undefined, message: string): void {
```
Comment (kennsippell) [inline src/worker/move-contact-worker.ts:185]:
```suggestion
  private static logWithTimestamp(job: Job|MinimalJob|undefined, message: string): void {
```

Sufficient?

File: src/worker/move-contact-worker.ts [hunk 5]
```diff
@@ -40,40 +46,56 @@ export class MoveContactWorker {
...
     console.log(`Job completed successfully: ${job.id}`);
     return true;
   };
 
-  private static async shouldPostpone(jobData: MoveContactData): Promise<boolean> {
+  private static handleRetryBackoff = (
+    attemptsMade: number, type: string | undefined, err: Error | undefined, job: MinimalJob | undefined
+  ): number => {
+    const {retryTimeFormatted} = this.computeRetryTime();
+    const fullMessage = `Job ${job?.id} will be retried ${attemptsMade + 1} time at ${retryTimeFormatted}. Due to failure: ${type}: ${err?.message}`;
```
Comment (kennsippell) [inline src/worker/move-contact-worker.ts:71]:
```suggestion
    const fullMessage = `Job ${job?.id} will retry at ${retryTimeFormatted}. Attempt Number: ${attemptsMade + 1}. Due to failure: ${type}: ${err?.message} `;
```

File: src/worker/move-contact-worker.ts [hunk 6]
```diff
@@ -40,40 +46,56 @@ export class MoveContactWorker {
...
       const errorMessage = err.response?.data?.error?.message || err.response?.error || err?.message;
       console.error('Error fetching monitoring data:', errorMessage);
-      return true;
+
+      // Handle server unavailability (HTTP 500 errors)
+      if (err.response?.status === 500) {
+        console.log('Server error encountered, postponing job...');
+        return { shouldPostpone: true, reason: `Server error encountered: ${errorMessage}` };
+      }
+      return { shouldPostpone: false, reason: '' };
```
Comment (kennsippell) [inline src/worker/move-contact-worker.ts:92]: This seems a bit odd here. Would it be clearer if this function returned `Promise<PostponeReason>` (or alike) and returned undefined when postponing is not required?

**Review by kennsippell — APPROVED** [review]
Nice

**Comment by kennsippell** [comment]
Does this fix Token Expiration?

File: src/liquid/place/move_form.html [hunk 7]
```diff
@@ -54,7 +54,12 @@ <h1 class="subtitle">To</h1>
 <script type="text/javascript">
   function handleResponse(event) {
     const response = event?.detail?.xhr?.responseText;
-    if (response && response.includes('data-success="true"')) {
+    const parser = new DOMParser();
+    const doc = parser.parseFromString(response, 'text/html');
+
+    // Find the element that indicates success
+    const successElement = doc.querySelector('[data-success="true"]');
+    if (successElement) {
```
Comment (kennsippell) [inline src/liquid/place/move_form.html:62]: Should something happen when !successElement?

Comment (paulpascal) [reply src/liquid/place/move_form.html:62]: Yeah, when no success (no data-success="true"), there is already the error block displaying an error notification...

```
  {% if error %}
    <div class="notification is-danger" style="margin: 0;">
      {{ error }}
    </div>
  {% endif %}

  {% if success %}
    <div data-success="true"></div>
  {% endif %}
  ```

**Comment by paulpascal** [comment]
> Does this fix Token Expiration? Does this fix Timeout Handling?

This is the only challenge remaining so far, I am thinking on two way to fix this:

- Increase the session token expiration (currently at 48h) 
- Dynamically refreshing the token (especially when the job is postponed)

Would like to have you though on that.

cc: @kennsippell @freddieptf

**Comment by kennsippell** [comment]
Recommend to increase the session token expiration (currently at 48h). This is straightforward. The tokens given during CHT login never expire. Can do in different PR if you desire it.

**Comment by paulpascal** [comment]
> Recommend to increase the session token expiration (currently at 48h). This is straightforward. The tokens given during CHT login never expire. Can do in different PR if you desire it.

Sure, I handled that in this [PR](https://github.com/medic/cht-user-management/pull/230)

**Pull request merged**